### PR TITLE
consensus/core: update sequence while a newChainHead event comes

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: consensus/pbft/core/new_chain_head.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		c.completed = true
+		c.setState(StateAcceptRequest)
+		// this block is from geth sync
+	} else {
+		logger.Debug("handleNewChainHead from geth sync", "height", ev.BlockNumber, "hash", ev.BlockHash)
+		// reset view number to 0
+		c.viewNumber = common.Big0
+		c.sequence = ev.BlockNumber
+		c.completed = true
+		if !c.setState(StateAcceptRequest) {
```
Comment (markya0616) [inline consensus/pbft/core/new_chain_head.go]: add force parameter

File: consensus/pbft/core/handler.go [hunk 2]
```diff
@@ -65,16 +61,8 @@ func (c *core) handleExternalEvent() {
...
-			// TODO: might have to handle error
-			c.sendCheckpoint(&pbft.Subject{
-				View: &pbft.View{
-					Sequence:   ev.BlockNumber,
-					ViewNumber: new(big.Int),
-				},
-				Digest: ev.BlockHash.Bytes(),
-			})
+		case pbft.NewChainHeadEvent:
+			c.handleNewChainHead(ev, c.backend.Validators().GetByAddress(c.address))
```
Comment (bailantaotao) [inline consensus/pbft/core/handler.go]: 使用c.Address()較好？

Comment (markya0616) [inline consensus/pbft/core/handler.go]: OK

File: consensus/pbft/core/new_chain_head.go (hunk 1 above)
Comment (markya0616) [inline consensus/pbft/core/new_chain_head.go]: force processBacklog in setState()

File: consensus/pbft/events.go [hunk 3]
```diff
@@ -41,7 +41,7 @@ type MessageEvent struct {
 	Payload []byte
 }
 
-type CheckpointEvent struct {
+type NewChainHeadEvent struct {
```
Comment (tailingchen) [inline consensus/pbft/events.go]: How about `FinalCommittedEvent` ?  I think `NewChainHead` is not the meaningful keyword in pbft.

Comment (markya0616) [inline consensus/pbft/events.go]: OK

File: consensus/pbft/core/handler.go [hunk 4]
```diff
@@ -65,22 +61,14 @@ func (c *core) handleExternalEvent() {
...
-			// TODO: might have to handle error
-			c.sendCheckpoint(&pbft.Subject{
-				View: &pbft.View{
-					Sequence:   ev.BlockNumber,
-					ViewNumber: new(big.Int),
-				},
-				Digest: ev.BlockHash.Bytes(),
-			})
+		case pbft.FinalCommittedEvent:
+			c.handleNewChainHead(ev, c.backend.Validators().GetByAddress(c.Address()))
```
Comment (tailingchen) [inline consensus/pbft/core/handler.go]: replace `NewChainHead` here.

Comment (markya0616) [inline consensus/pbft/core/handler.go]: 對耶XDDD

File: consensus/pbft/core/final_committed.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+				ViewNumber: c.viewNumber,
+			},
+			Digest: ev.BlockHash.Bytes(),
+		})
+		c.snapshotsMu.Lock()
+		c.snapshots = append(c.snapshots, c.current)
+		c.snapshotsMu.Unlock()
+
+		// reset view number to 0
+		c.viewNumber = common.Big0
```
Comment (tailingchen) [inline consensus/pbft/core/final_committed.go]: It's better to keep the current view number here. we might need the view number to compare the priority of future message.
Use `new(big.Int).Set()` to copy the value

File: consensus/pbft/core/final_committed.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+		c.sequence = c.current.Sequence
+		c.completed = true
+		c.setState(StateAcceptRequest)
+		// this block is from geth sync
+	} else {
+		logger.Debug("handleFinalCommitted from geth sync", "height", ev.BlockNumber, "hash", ev.BlockHash)
+		// reset view number to 0
+		c.viewNumber = common.Big0
+		c.sequence = ev.BlockNumber
```
Comment (tailingchen) [inline consensus/pbft/core/final_committed.go]: Use new(big.Int).Set() to copy the value.

File: consensus/pbft/core/final_committed.go (hunk 5 above)
Comment (markya0616) [inline consensus/pbft/core/final_committed.go]: ＯＫ

File: consensus/pbft/core/final_committed.go (hunk 6 above)
Comment (markya0616) [inline consensus/pbft/core/final_committed.go]: OK

File: consensus/pbft/core/final_committed.go (hunk 5 above)
Comment (markya0616) [inline consensus/pbft/core/final_committed.go]: ok

File: consensus/pbft/core/final_committed.go (hunk 6 above)
Comment (markya0616) [inline consensus/pbft/core/final_committed.go]: ok

**Pull request merged**